client/src/visualiser-debugger/Component/Configuration/StackVarDeclaration.tsx

```diff
@@ -19,10 +19,32 @@ export const StackVarAnnotation: React.FC<StackVariableAnnotationProp> = ({
   name,
   memoryValue,
 }: StackVariableAnnotationProp) => {
-  const [selectedRole, setSelectedRole] = useState<StackVariableRole>(StackVariableRole.Empty);
+  // Annotate by default if the variable contains a pointer
+  const [selectedRole, setSelectedRole] = useState<StackVariableRole>(
+    isPointerType(memoryValue.typeName)
+      ? StackVariableRole.LinkedListPointer
+      : StackVariableRole.Empty
+  );
   const [isDropdownOpen, setIsDropdownOpen] = useState(false);
-  const { userAnnotation } = useGlobalStore().visualizer;
-  const { updateUserAnnotation } = useGlobalStore();
+  const stackAnnotation = useGlobalStore(
+    (state) => state.visualizer.userAnnotation.stackAnnotation
+  );
+  const updateStackAnnotation = useGlobalStore((state) => state.updateStackAnnotation);
+
+  // Annotate by default if the variable contains a pointer
+  useEffect(() => {
+    if (selectedRole === StackVariableRole.LinkedListPointer) {
+      const newStackAnnotation = {
+        [name]: memoryValue.typeName,
+      };
+      updateStackAnnotation(newStackAnnotation);
+    }
+
+    // Persist the stack variable state
+    if (name in stackAnnotation && stackAnnotation[name] === null) {
+      setSelectedRole(StackVariableRole.Empty);
+    }
+  }, []);
```

Comment:
What if it is a pointer but not to a linked list node struct. It is still automatically classified as a `StackVariableRole.LinkedListPointer`? Maybe should extra check for whether this stack var's type is a linked list type. See `createPossibleLinkedListTypeDecl()` in LinkedListAnnotion.tsx`  and ask me if it is hard to understand

Reply:
Do you mean instead of just `isPointerType(memoryValue.typeName)`, I also need to check `isStructTypeName(memoryValue.typeName.slice(0, -1))`?

So, it'll look like this
`const [selectedRole, setSelectedRole] = useState<StackVariableRole>(
    isPointerType(memoryValue.typeName) && isStructTypeName(memoryValue.typeName.slice(0, -1))
      ? StackVariableRole.LinkedListPointer
      : StackVariableRole.Empty
  );`

Reply:
Yea thats right :D

Reply:
What you have written will check that the type is a struct pointer. However we can do more.. we can check whether the type is a valid linked list type. See how `createPossibleLinkedListTypeDecl()` does it using a `BackendTypeDeclaration`:

```js
export type BackendTypeDeclaration = {
  file: string;
  line_num: string;
  original_line: string;
  typeName: NativeTypeName;
  // For structs, the fields
  fields?: {
    name: Name;
    typeName: NativeTypeName;
  }[];
  // For typedefs, the type being aliased
  type?: {
    typeName: NativeTypeName;
  };
};
```

Example of a `BackendTypeDeclaration` that is a valid linked list node type:

```json
{
  ...
  typeName: "struct city",
  fields: [
    {
      name: "population";
      typeName: "int";
    },
    {
      name: "nextCity";
      typeName: "struct city *";
    },
  ],
}

Now suppose we pass the above `BackednTypeDeclaration` to the function `createPossibleLinkedListTypeDecl()`, which will try to figure out whether it is a valid linked list node type.

Criteria for determining whether it is a valid linked list node type:
- has some kind of "value" field e.g `int population` (can be anything really)
- has a field that is a pointer to the struct itself e.g. `struct city *`

Then the `possibleTypeDecl` created and returned by the function will be:

```json
{
  typeName: "struct city",
  possibleValues: [
    {
      name: "population",
      typeName: "int",
    },
  ],
  possibleNexts: [
    {
      name: "nextCity",
      typeName: "struct city*",
    },
  ],
}
````

```js
export const createPossibleLinkedListTypeDecl = (
  typeDecl: BackendTypeDeclaration
): PossibleLinkedListAnnotation | null => {
  if (!('fields' in typeDecl)) {
    return null;
  }
  if (!isStructTypeName(typeDecl.typeName)) {
    return null;
  }

  const possibleTypeDecl: PossibleLinkedListAnnotation = {
    typeName: typeDecl.typeName,
    possibleValues: [],
    possibleNexts: [],
  };

  typeDecl.fields.forEach((field) => {
    if (isNativeTypeName(field.typeName)) {
      possibleTypeDecl.possibleValues.push({
        name: field.name,
        typeName: field.typeName,
      });
    }
    if (
      isPointerType(field.typeName) &&
      isStructTypeName(field.typeName.slice(0, -1)) &&
      field.typeName.includes(typeDecl.typeName)
    ) {
      possibleTypeDecl.possibleNexts.push({
        name: field.name,
        typeName: field.typeName,
      });
    }
  });

  if (possibleTypeDecl.possibleValues.length >= 1 && possibleTypeDecl.possibleNexts.length >= 1) {
    return possibleTypeDecl;
  }

  return null;
};
```

If the type passed in is not a valid linekd list node type, then `createPossibleLinkedLIstTypeDecl()` returns `null`

So you could use this function to determine whether a stack variable is a valid linked list or not.